### PR TITLE
[slurm] Add slurm plugin

### PR DESCRIPTION
**Pull request by arif-ali**
Slurm is a workload manager in the HPC space, this is a start on this, and there may be further additions in the future

Resolves: #3329

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

### CODE REVIEW

File: sos/report/plugins/slurm.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, UbuntuPlugin, RedHatPlugin
+
+
+class Slurm(Plugin, UbuntuPlugin, RedHatPlugin):
+
```
Comment (pmoravec) [inline sos/report/plugins/slurm.py:13]: Please add `short_desc` (the reason for tests failing :), in final PR version ).

File: sos/report/plugins/slurm.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        'slurmd',
+        'slurmdbd',
+        'slurmctld',
+    )
+
+    def setup(self):
+        """ Slurm Workload Manager
+        """
+
+        self.add_copy_spec('/etc/slurm/')
```
Comment (pmoravec) [inline sos/report/plugins/slurm.py:38]: Can't be there some credentials? Reading https://slurm.schedmd.com/slurm.conf.html, `AccountingStoragePass` or `JobCompPass` seems storing passwords..?

**Review by TurboTurtle — APPROVED** [review]

File: sos/report/plugins/slurm.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+        self.add_copy_spec('/etc/slurm/*.conf')
+
+        if is_executable('sinfo'):
+            self.add_cmd_output([
+                'sinfo --all --list-reasons --long',
+                'sinfo --all --long',
+            ])
+
+        if is_executable('squeue'):
```
Comment (pmoravec) [inline sos/report/plugins/slurm.py:49]: I think this test might not be needed. If a command isn't executable, `sos` won't store output of it (https://github.com/sosreport/sos/blob/main/sos/utilities.py#L261-L270) - but will add a record to `manifest.json`.

But definitely it is fine to use the test here.

File: sos/report/plugins/slurm.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            'licenses',
+            'node',
+            'partition',
+            'reservation',
+            'slurmd',
+            'step',
+            'topology',
+        ]
+
+        if is_executable('scontrol'):
```
Comment (pmoravec) [inline sos/report/plugins/slurm.py:74]: Yeah, here the test makes a lot of sense :) (test executable once, save many "command not found" deadends ahead). :)

**Review by pmoravec — APPROVED** [review]

**Comment by arif-ali** [comment]
I tested it without the test, and the file was created, so hence added the tests.

```
# more sosreport-slurm-comp01-2023-08-08-ddtiwod/sos_commands/slurm/squeue_--all_--long
timeout: failed to run command ‘squeue’: No such file or directory
```

I could easily split the plugin out to multiple plugins, one for `slurmd`, `slurmctld` and `slurmdbd`, but pointless imho. It was easier to add everything in one rather than multiple plugins.

`slurmd` nodes (on their own) will never have the commands, but logs would be there, the commands are only on hosts that the server service is running, `slurm` or EL hosts, and `slurm-client` for Ubuntu.

Alternative method would have been to test for the above 2 packages for the command availability

**Comment by pmoravec** [comment]
> I tested it without the test, and the file was created, so hence added the tests.
> 
> ```
> # more sosreport-slurm-comp01-2023-08-08-ddtiwod/sos_commands/slurm/squeue_--all_--long
> timeout: failed to run command ‘squeue’: No such file or directory
> ```
> 
> I could easily split the plugin out to multiple plugins, one for `slurmd`, `slurmctld` and `slurmdbd`, but pointless imho. It was easier to add everything in one rather than multiple plugins.
> 
> `slurmd` nodes (on their own) will never have the commands, but logs would be there, the commands are only on hosts that the server service is running, `slurm` or EL hosts, and `slurm-client` for Ubuntu.
> 
> Alternative method would have been to test for the above 2 packages for the command availability

Ah I see, then the test makes a lot of sense on either place.

If there is more stuff to collect common than different, I would stay with one plugin.

**Pull request merged**